### ISSUE TITLE
How pySHACL validates classes with parents seems to be wrong

### ISSUE THREAD

**Issue by vemonet** [RDFLib/pySHACL#173]
Hi, while using pySHACL to validate RDF complying with the [BioLink model](https://github.com/biolink/biolink-model) I noticed that how pySHACL handles validation of parents classes of a class is wrong. 

Because pySHACL tries to validate an entity complies with the shapes of all its parent classes.

## Reproduce

Install dependencies:

```bash
pip install pyshacl rdflib
```

Run this script:

```python
import pyshacl
from rdflib import Graph

validate_rdf = {
    "@context": {
        "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
        "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
        "dct": "http://purl.org/dc/terms/",
        "biolink": "https://w3id.org/biolink/vocab/"
    },
    "@type": "biolink:Drug",
    "biolink:category": "biolink:Drug",
    "biolink:id": "drugbank:DB00001",
    "rdfs:label": "Lepirudin",
    "dct:description": "Lepirudin is a protein-based direct thrombin inhibitor."
}

validate_g = Graph()
validate_g.parse(
    data=validate_rdf,
    format="ttl"
)

shacl_g = Graph()
shacl_g.parse(
    "https://raw.githubusercontent.com/vemonet/biolink-model/add-shacl-gen/biolink-model.shacl.ttl",
    format="ttl"
)

biolink_g = Graph()
biolink_g.parse(
    "https://raw.githubusercontent.com/biolink/biolink-model/master/biolink-model.owl.ttl",
    format="ttl"
)

conforms, _, results_text = pyshacl.validate(
    validate_g, 
    shacl_graph=shacl_g,
    ont_graph=biolink_g
)

print(results_text)
```

> I tested it with python 3.9.15 on Fedora 37

You will get **23** errors

If you look at those errors, all of them are `ClosedConstraintComponent`, and the source shapes are all parents of the `biolink:Drug` shape.

## What happens

By adding the BioLink ontology it adds the subclassOf hierarchy. This leads pySHACL to try to evaluate a node against the shapes of all its parent classes. This leads to misleading errors, especially if the parent shapes are closed.

For example, I want to validate an entity is a `biolink:Drug`, currently pySHACL tries to validate that my entity complies with the shapes of every parent classes (e.g. `biolink:ChemicalSubstance` and `biolink:NamedThing`).

But it does not seems to be correct, because my Drug might have additional properties that are not described in `biolink:ChemicalSubstance` and `biolink:NamedThing`. And pySHACL will always fail validating such valid classes right now 

## How to fix it

I managed to add a quick fix for this behavior here by commenting the part that was adding parent classes shapes to a node target: https://github.com/vemonet/pySHACL/blob/e51c658bba4dee896e54be243d191f7dd8b321c0/pyshacl/shape.py#L343:L350 

## Questions

Note that I don't know much about how pySHACL works, so I might be missing something, maybe there is a valid reason behind those checks, and the real issue comes from somewhere else. What is this reason? Why do we need these checks? 

What could be a good solution to move forward? I can send a PR if you are interested by those changes already

@ashleysommer @ajnelson-nist

**Comment by ashleysommer** [comment]
Hi @vemonet 
I've taken some time to look into this issue that you've described. 

Let me first take a moment to check that I correctly understand the issue you are facing:

```
#shapes.ttl
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org#> .

ex:AnimalShape rdf:type sh:NodeShape ;
  sh:targetClass ex:Animal ;
  sh:property [
      sh:path ex:name;
    ];
  sh:property [
      sh:path ex:legsCount;
      sh:minInclusive 0;
    ];
  sh:closed true;
  sh:ignoredProperties ( rdf:type )
  .

ex:BirdShape rdf:type sh:NodeShape ;
  sh:targetClass ex:Bird ;
  sh:property [
    sh:path ex:name;
  ];
  sh:property [
      sh:path ex:wingsCount;
      sh:minInclusive 0;
    ];
  sh:property [
      sh:path ex:legsCount;
      sh:minInclusive 0;
    ];
  sh:closed true;
  sh:ignoredProperties ( rdf:type )
  .
```

```
#data.ttl
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org#> .
@prefix data: <http://data.org#> .

ex:Animal rdf:type rdfs:Class .
ex:Bird rdf:type rdfs:Class .
ex:Bird rdfs:subClassOf ex:Animal .

data:Cheetah;
  rdf:type ex:Animal ;
  ex:name "Cheetah";
  ex:legsCount 4;
 .

data:Eagle;
  rdf:type ex:Bird ;
  ex:name "Eagle";
  ex:legsCount 2;
  ex:wingsCount 2;
 .
```
_I have included the RDFS typing information in the data graph, to avoid needing three graphs in this example._

I am assuming you are running with no RDFS or OWL pre-inferencing enabled in the validator, because the issue you described only occurs when you are expecting no inferencing associations to be present.

When I run validator on the example above, I get the following validation report:
```
Constraint Violation in ClosedConstraintComponent (http://www.w3.org/ns/shacl#ClosedConstraintComponent):
	Severity: sh:Violation
	Source Shape: ex:AnimalShape
	Focus Node: data:Eagle
	Value Node: Literal("2", datatype=xsd:integer)
	Result Path: ex:wingsCount
	Message: Node data:Eagle is closed. It cannot have value: Literal("2", datatype=xsd:integer)
```

Correct me if I'm wrong, as I understand it you are asserting that this example should pass validation. You are stating that the entity `data:Eagle` should _not_ be a focus node of the NodeShape `ex:AnimalShape`, is that right?

Indeed, commenting out that section of code you mentioned, it skips adding subclasses of ex:Animal to its list of focus nodes for `ex:AnimalShape`, and this example does then validate successfully.

I'm sorry to conclude in this case the first result is true, and PySHACL is acting correctly. By commenting out that section of code, PySHACL no longer passes its internal W3C conformance tests, and would no longer be considered a compliant SHACL validator implementation.

The [definition in the W3C SHACL Specification document for the implementation of `sh:targetClass` and implicit class targeting is quite clear](https://www.w3.org/TR/shacl/#targetClass). It states: 
> If **s** is a shape in a shapes graph **SG** and **s** has [value](https://www.w3.org/TR/shacl/#dfn-value) **c** for `sh:targetClass` in **SG** then the set of [SHACL instances](https://www.w3.org/TR/shacl/#dfn-shacl-instance) of **c** in a data graph **DG** is a [target](https://www.w3.org/TR/shacl/#dfn-target) from **DG** for **s** in **SG**.

That is hard to follow, but the important part is the definition of [SHACL instances](https://www.w3.org/TR/shacl/#dfn-shacl-instance):
> **SHACL Class Instance**
>A [node](https://www.w3.org/TR/shacl/#dfn-node) **n** in an [RDF graph](https://www.w3.org/TR/shacl/#dfn-rdf-graph) **G** is a SHACL instance of a [SHACL class](https://www.w3.org/TR/shacl/#dfn-shacl-class) **C** in **G** if one of the [SHACL types](https://www.w3.org/TR/shacl/#dfn-shacl-types) of **n** in **G** is **C**.

And the definition of [SHACL types](https://www.w3.org/TR/shacl/#dfn-shacl-types):
>**SHACL Type**
> The SHACL types of an [RDF term](https://www.w3.org/TR/shacl/#dfn-rdf-term) in an [RDF graph](https://www.w3.org/TR/shacl/#dfn-rdf-graph) is the set of its [values](https://www.w3.org/TR/shacl/#dfn-value) for rdf:type in the [graph](https://www.w3.org/TR/shacl/#dfn-rdf-graph) as well as the [SHACL superclasses](https://www.w3.org/TR/shacl/#dfn-shacl-superclass) of these [values](https://www.w3.org/TR/shacl/#dfn-value) in the [graph](https://www.w3.org/TR/shacl/#dfn-rdf-graph).

And finally, the definition for [SHACL superclasses](https://www.w3.org/TR/shacl/#dfn-shacl-superclass):
> SHACL Subclass, SHACL superclass
> A [node](https://www.w3.org/TR/shacl/#dfn-node) **Sub** in an [RDF graph](https://www.w3.org/TR/shacl/#dfn-rdf-graph) is a _SHACL subclass_ of another [node](https://www.w3.org/TR/shacl/#dfn-node) **Super** in the [graph](https://www.w3.org/TR/shacl/#dfn-rdf-graph) if there is a sequence of [triples](https://www.w3.org/TR/shacl/#dfn-rdf-triple) in the [graph](https://www.w3.org/TR/shacl/#dfn-rdf-graph) each with predicate `rdfs:subClassOf` such that the [subject](https://www.w3.org/TR/shacl/#dfn-subject) of the first [triple](https://www.w3.org/TR/shacl/#dfn-rdf-triple) is **Sub**, the [object](https://www.w3.org/TR/shacl/#dfn-object) of the last triple is **Super**, and the [object](https://www.w3.org/TR/shacl/#dfn-object) of each [triple](https://www.w3.org/TR/shacl/#dfn-rdf-triple) except the last is the [subject](https://www.w3.org/TR/shacl/#dfn-subject) of the next. If **Sub** is a [SHACL subclass](https://www.w3.org/TR/shacl/#dfn-shacl-subclass) of **Super** in an [RDF graph](https://www.w3.org/TR/shacl/#dfn-rdf-graph) then **Super** is a _SHACL superclass_ of **Sub** in the [graph](https://www.w3.org/TR/shacl/#dfn-rdf-graph).

The part of the code you have commented out is exactly the part that implements the subclass/superclass directives in the targeting specification described above, and removing it breaks compatibility with the specification.

If I have misunderstood the issue, please let me know, and we can explore it further.

If you still disagree with the intended behaviour, then it is no longer an issue with PySHACL, rather an issue with the SHACL standard itself. Either way, the best places to discuss this topic further are the [SHACL W3C mailing list](https://www.w3.org/community/shacl/), and [SHACL community Discord server](https://discord.gg/RTbGfJqdKB), folks on there will be able to give you much better advice than I can.

**Comment by ajnelson-nist** [comment]
Hi @vemonet ,

The issue I see in the SHACL shapes is usage of `sh:closed`, which might have a farther-reaching (-constricting) effect than you may realize.

Per the [SHACL spec, 4.8.1](https://www.w3.org/TR/shacl/#ClosedConstraintComponent):

> The SHACL Core language includes a property called sh:closed that can be used to specify the condition that each value node has [values](https://www.w3.org/TR/shacl/#dfn-value) only for those properties that have been explicitly enumerated via the [property shapes](https://www.w3.org/TR/shacl/#dfn-property-shape) specified for the shape via `sh:property`.

Re-iterating: `sh:closed` disallows *all* extensibility from properties not mentioned via `sh:property`.  `rdf:type` seems to get a pass for bootstrapping reasons (SHACL needs to know types for targeting purposes), but if there was no `sh:PropertyShape` for, say, `rdfs:comment`, your SHACL validation would fail if you put an `rdfs:comment` on the individual-node you're trying to validate.

So, this is going to be a conflict with your remark in your problem description:

> my Drug might have additional properties that are not described in `biolink:ChemicalSubstance` and `biolink:NamedThing`.

`biolink:NamedThing` has `sh:closed` set to `true`.  This unfortunately hampers making subclasses of `biolink:NamedThing` that will want to apply their own properties.

So, in my opinion, your shapes can't work with use of `sh:closed`.  I haven't seen the documentation for this ontology, so I don't know why you were motivated to try `sh:closed`.  My community had a [similar desire](https://github.com/casework/CASE-Utilities-Python/issues/40) for concept IRI typo-checking *within our community namespaces*, but had to resort to writing a piece of extension functionality in a `pyshacl` wrapper.  What we wanted probably would have been satisfied if SHACL had a mechanism for something like `sh:closed`, but just scoped to a certain prefix.
